junorunner/runner.py

```diff
@@ -17,108 +22,100 @@ class JunoDiscoverRunner(DiscoverRunner):
-                kwargs['top_level_dir'] = top_level
-
-            if not (tests and tests.countTestCases()):
-                # if no tests found, it's probably a package; try discovery
-                tests = self.test_loader.discover(start_dir=label, **kwargs)
-
-                # make unittest forget the top-level dir it calculated from this
-                # run, to support running tests from two different top-levels.
-                self.test_loader._top_level_dir = None
-
-            tests = self.get_tests_defined_in_methods_or_none(tests, methods)
-            if tests:
-                suite.addTests(tests)
-
-        for test in extra_tests:
-            suite.addTest(test)
-        return suite
-
-    def _get_parallel_suite(self, suite):
-        if self.parallel > 1:
-            parallel_suite = self.parallel_test_suite(suite, self.parallel, self.failfast)
-
-            # Since tests are distributed across processes on a per-TestCase
-            # basis, there's no need for more processes than TestCases.
-            parallel_units = len(parallel_suite.subsuites)
-            if self.parallel > parallel_units:
-                self.parallel = parallel_units
-
-            # If there's only one TestCase, parallelization isn't needed.
-            if self.parallel > 1:
-                return parallel_suite
-        return suite
+    def get_test_runner_kwargs(self):
+        return {
+            "failfast": self.failfast,
+            "resultclass": self.get_resultclass(),
+            "verbosity": self.verbosity,
+            "buffer": self.buffer,
+            "total_tests": self.total_tests,
+            "slow_test_count": self.slow_test_count,
+            "use_log_files": self.use_log_files,
+        }
 
     def build_suite(self, test_labels=None, extra_tests=None, **kwargs):
-        extra_tests = extra_tests or []
-
         methods = self.methods.split(',') if self.methods else []
         if methods:
             self.use_log_files = False
 
+        if extra_tests is not None:
+            warnings.warn(
+                "The extra_tests argument is deprecated.",
+                RemovedInDjango50Warning,
+                stacklevel=2,
+            )
+        test_labels = test_labels or ["."]
+        extra_tests = extra_tests or []
+
         discover_kwargs = {}
         if self.pattern is not None:
-            discover_kwargs['pattern'] = self.pattern
+            discover_kwargs["pattern"] = self.pattern
         if self.top_level is not None:
-            discover_kwargs['top_level_dir'] = self.top_level
+            discover_kwargs["top_level_dir"] = self.top_level
+        self.setup_shuffler()
+
+        all_tests = []
+        for label in test_labels:
+            tests = self.load_tests_for_label(label, discover_kwargs)
+            tests = self.get_tests_defined_in_methods_or_none(tests, methods)
+            if tests:
+                all_tests.extend(iter_test_cases(tests))
```

Comment:
This is the only part changed of this method, the rest is copy/paste from django.